src/wasm/wasm-type.cpp

```diff
@@ -858,9 +858,9 @@ bool TypeComparator::lessThan(HeapType a, HeapType b) {
     return false;
   }
   if (seen.count({a, b})) {
-    // We weren't able to disprove that a < b since we last saw them, so the
-    // relation holds coinductively.
-    return true;
+    // We weren't able to disprove that a == b since we last saw them, so it
+    // holds coinductively. a < b is therefore false;
+    return false;
```

Comment:
It's not obvious to me why this is enough, but maybe I don't know the C++ spec well enough. Do you have a link?

My intuition is that before you could have `a < b, b < a, a != b`, while now you will have `a >= b, b >= a, a != b` which from a logical point of view is equivalent.

Reply:
https://en.cppreference.com/w/cpp/named_req/Compare. The key point is that it _only_ requires < to create a weak ordering and it's ok if == is not consistent with that ordering.

Reply:
I see, thanks! So yes, it must be a strict weak ordering, and so `!(a < b), !(b < a), a != b` is fine.